pallets/alt-benchmarks/src/bench_parachains_slashing.rs

```diff
@@ -0,0 +1,186 @@
+    let session_info = session_info.unwrap();
+    let validator_id = session_info
+        .validators
+        .get(ValidatorIndex::from(0))
+        .unwrap()
+        .clone();
+    let key = (PARACHAIN_KEY_TYPE_ID, validator_id.clone());
+    let key_owner_proof = pallet_session::historical::Pallet::<T>::prove(key).unwrap();
+
+    // rotate a session to make sure `key_owner_proof` is historical
+    initializer::Pallet::<T>::on_initialize(BlockNumberFor::<T>::one());
+    pallet_session::Pallet::<T>::rotate_session();
+    initializer::Pallet::<T>::on_finalize(BlockNumberFor::<T>::one());
+
+    let idx = polkadot_runtime_parachains::shared::CurrentSessionIndex::<T>::get();
+    assert!(
+        idx > session_index,
+        "session rotation should work for parachain pallets: {} <= {}",
+        idx,
+        session_index,
+    );
+
+    (session_index, key_owner_proof, validator_id)
+}
+
+/// Submits a single `ForInvalid` dispute.
+fn setup_dispute<T>(session_index: SessionIndex, validator_id: ValidatorId) -> DisputeProof
+where
+    T: Config,
+{
+    let current_session = T::ValidatorSet::session_index();
+    assert_ne!(session_index, current_session);
+
+    let validator_index = ValidatorIndex(0);
+    let losers = [validator_index].into_iter();
+    let backers = losers.clone();
+
+    T::SlashingHandler::punish_for_invalid(session_index, CANDIDATE_HASH, losers, backers);
+
+    // // TODO: Storage is pub(crate) so cannot be accessed here
+    // let unapplied = <UnappliedSlashes<T>>::get(session_index, CANDIDATE_HASH);
+    // assert_eq!(unapplied.unwrap().keys.len(), 1);
+
+    dispute_proof(session_index, validator_id, validator_index)
+}
+
+/// Creates a `ForInvalid` dispute proof.
+fn dispute_proof(
+    session_index: SessionIndex,
+    validator_id: ValidatorId,
+    validator_index: ValidatorIndex,
+) -> DisputeProof {
+    let kind = SlashingOffenceKind::ForInvalid;
+    let time_slot = DisputesTimeSlot::new(session_index, CANDIDATE_HASH);
+
+    DisputeProof {
+        time_slot,
+        kind,
+        validator_index,
+        validator_id,
+    }
+}
+
+#[benchmarks(where T: Config<KeyOwnerProof = MembershipProof>)]
+mod benchmarks {
+    use super::*;
+
+    #[benchmark]
+    fn report_dispute_lost_unsigned(n: Linear<4, MAX_VALIDATORS>) {
+        log::info!("setup_validator_set");
+        let (session_index, key_owner_proof, validator_id) = setup_validator_set::<T>(n);
+
+        log::info!("setup_dispute");
+        // submit a single `ForInvalid` dispute for a past session.
+        let dispute_proof = setup_dispute::<T>(session_index, validator_id);
+
+        log::info!("extrinsic call");
+        #[extrinsic_call]
+        _(RawOrigin::None, Box::new(dispute_proof), key_owner_proof);
+
+        // // TODO: Storage is pub(crate) so cannot be accessed here
+        // let unapplied = <UnappliedSlashes<T>>::get(session_index, CANDIDATE_HASH);
```

Comment:
which storage is this? is it ours? if so just make it public

Reply:
It is part of `parachains_slashing` which is part of PolkadotSDK.